rust/src/normalization.rs

```diff
@@ -120,6 +121,294 @@ pub(crate) fn rewrite_rpmdb_timestamps<F: Read + Write + Seek>(rpmdb: &mut F) ->
+                        ))?;
+
+                        if matches!(header.page_type, PageType::IBTree | PageType::LBTree) {
+                            let item: BTreeItem = db.read_le()?;
+                            if header.page_type == PageType::IBTree {
+                                db.write_all(b"\x00")?;
+                            } else if header.page_type == PageType::LBTree {
+                                if let BTreeItemType::Overflow = item.item_type {
+                                    db.seek(SeekFrom::Current(-3))?;
+                                    db.write_all(b"\x00\x00")?;
+                                } else if let BTreeItemType::KeyData = item.item_type {
+                                    let next_offset = if let Some(next) = offset_iter.peek() {
+                                        *next
+                                    } else {
+                                        meta_header.pagesize as u16
+                                    };
+                                    let remainder = next_offset - (offset + 3 + item.len);
+                                    if remainder != 0 {
+                                        db.seek(SeekFrom::Current(item.len as i64))?;
+                                        std::io::copy(
+                                            &mut std::io::repeat(b'\x00').take(remainder as u64),
+                                            &mut db,
+                                        )?;
+                                    }
+                                }
+                            }
+                        } else if header.page_type == PageType::Hash {
+                            let item_type: HashItemType = db.read_le()?;
+
+                            if let HashItemType::Offpage = item_type {
+                                db.write_all(b"\x00\x00\x00")?;
+                            }
+                        }
+                    }
+                }
+
+                db.flush()?;
+            }
+
+            let r = Exec::cmd("db_verify")
+                .args(&[OsStr::new("-q"), PROC_SELF_CWD.join(&path).as_os_str()])
+                .cwd(PROC_SELF_FD.join(rootfs.as_raw_fd().to_string()))
+                .join()?;
+            if !r.success() {
+                return Err(anyhow!("post-normalization verification failed"));
+            }
+
+            let new_digest = sha2::Sha256::digest(
+                Exec::cmd("db_dump")
+                    .args(&[PROC_SELF_CWD.join(&path)])
+                    .cwd(PROC_SELF_FD.join(rootfs.as_raw_fd().to_string()))
+                    .stdout(Redirection::Pipe)
+                    .capture()
+                    .context("post-normalization data dump")?
+                    .stdout
+                    .as_slice(),
+            );
+
+            if new_digest != old_digest {
+                return Err(anyhow!("file data changed during normalization"));
```

Comment:
Let's include something mentioning "bdb" in this string?  Maybe:
`Err(anyhow!("bdb normalization failed, detected content change"))` ?